aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp

```diff
@@ -563,6 +422,280 @@ void upsample_trilinear3d_backward_kernel_impl(
+  //
+  // Indices dtype can be int32_t or int64_t depending on canUse32BitIndexMath(input)
+  // which should not overflow because maximum possible value that it could take is the 
+  // product of interpolated input strides: input_size[dim-1] * input_size[dim-2] * ...
+  // which is always smaller then the number of input elements checked by canUse32BitIndexMath
+  std::vector<std::vector<Tensor>> indices_weights;
+  AT_DISPATCH_FLOATING_TYPES(
+    input.scalar_type(), "compute_indices_weights_linear", [&] {
+      auto es = input.element_size();
+      for (int i=0; i<out_ndims; i++) {
+        indices_weights.emplace_back(
+          ti_compute_indices_weights_linear<index_t, scalar_t>(
+            input.size(i + 2), oshape[i + 2], input.stride(i + 2) * es, input.dim(), i + 2, align_corners, scales[i])
+        );
+      }
+    }
+  );
+
+  TensorIteratorConfig config;
+  config.check_all_same_dtype(false)
+    .declare_static_dtype_and_device(input.scalar_type(), input.device())
+    .add_output(output)
+    .add_input(restrided_input);
+  
+  for (auto iter=indices_weights.begin(); iter!=indices_weights.end(); iter++) { 
+    for (auto& tensor : *iter) {
+      config.add_input(tensor);
+    }
+  }
+
+  auto iter = config.build();
+
+  AT_DISPATCH_FLOATING_TYPES(
+      iter.dtype(), "upsample_linearNd", [&] {
+      ti_cpu_upsample_linear<scalar_t, index_t, out_ndims>(iter);
+  });
+
+}
+
+using scale_t = std::vector<c10::optional<double>>;
+void ti_upsample_bilinear2d_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    c10::optional<double> scales_h,
+    c10::optional<double> scales_w) {
+
+  if (canUse32BitIndexMath(input)) {
+    ti_upsample_linearNd_kernel_impl<int32_t, 2, scale_t>(
+      output, input, align_corners, {scales_h, scales_w});
+  } else {
+    ti_upsample_linearNd_kernel_impl<int64_t, 2, scale_t>(
+      output, input, align_corners, {scales_h, scales_w});  
+  }
+}
+
+void upsample_linear1d_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    c10::optional<double> scales_w) {
+
+  if (canUse32BitIndexMath(input)){
+    ti_upsample_linearNd_kernel_impl<int32_t, 1, scale_t>(
+      output, input, align_corners, {scales_w});
+  } else {
+    ti_upsample_linearNd_kernel_impl<int64_t, 1, scale_t>(
+      output, input, align_corners, {scales_w});
+  }
+}
+
+void ti_upsample_trilinear3d_kernel_impl(
+    Tensor& output,
+    const Tensor& input,
+    bool align_corners,
+    c10::optional<double> scales_d,
+    c10::optional<double> scales_h,
+    c10::optional<double> scales_w) {
+
+  if (canUse32BitIndexMath(input)) {
+    ti_upsample_linearNd_kernel_impl<int32_t, 3, scale_t>(
+      output, input, align_corners, {scales_d, scales_h, scales_w});
+  } else {
+    ti_upsample_linearNd_kernel_impl<int64_t, 3, scale_t>(
+      output, input, align_corners, {scales_d, scales_h, scales_w});
+  }
+}
```

Comment:
Given that those are helper functions, can we maybe refactor this into a single function instead, with the same API as `cpu_upsample_linear`?